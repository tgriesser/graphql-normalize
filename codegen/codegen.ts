import type { CodegenPlugin } from '@graphql-codegen/plugin-helpers';

const NormalizeCodegenPlugin: CodegenPlugin = {
  plugin(schema, documents, config, info) {
    //
  },
};

export default NormalizeCodegenPlugin;
