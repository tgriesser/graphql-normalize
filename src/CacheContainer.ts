import produce, { Immutable } from 'immer';
import type { CacheShape } from './cache';

export class CacheContainer {
  #state: Immutable<CacheShape>;
  #subscribers: Function[] = [];

  constructor(initialState?: CacheShape) {
    this.#state = produce({ fields: {}, operations: {} }, (root) => initialState ?? root);
  }

  getState() {
    return this.#state;
  }

  subscribe(fn: (state: Immutable<CacheShape>) => void) {
    this.#subscribers.push(fn);
    return () => {
      this.#subscribers.splice(this.#subscribers.indexOf(fn), 1);
    };
  }
}
