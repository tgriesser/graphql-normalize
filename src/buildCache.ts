import type { FormattedExecutionResult } from 'graphql';
import { walk } from './walk';
import type { NormalizedDoc } from './extensionShape';

/**
 * Write values into the cache, given the values, the existing cache, and the
 */
export function buildCache(result: FormattedExecutionResult, normalizingMeta: NormalizedDoc) {
  let pathPrefix = '';
  let currentKey;
  let currentKeyStack = [];
  let currentWriteKey = undefined;
  let currentWriteKeyStack = [];

  const selectionKeys = Object.keys(normalizingMeta.selectionSet);

  function walkUntilList(val: any) {
    walk(val, {
      enter(node) {
        const key = node.path.join('.');
        const metaDef = normalizingMeta.selectionSet[key];
        console.log(key, metaDef);
        if (metaDef.list) {
          return false;
        }
      },
      leave() {
        //
      },
    });
  }

  walkUntilList(result);
}
