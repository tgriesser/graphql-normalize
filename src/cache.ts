import type { NormalizeMetaShape } from './metadataShapes';

export interface CacheShape {
  operations: {
    // The metadata for each query
    [hash: string]: {
      meta: NormalizeMetaShape;
    };
  };
  fields: {
    [key: string]: any;
  };
}
