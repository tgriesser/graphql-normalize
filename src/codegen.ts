import { NormalizeCodegenPlugin } from './codegen/plugin.js';
export type { TypePolicies } from './codegen/getCacheKey.js';
export type { NormalizeMetaShape, NormalizedDocShape } from './metadataShapes.js';

export { NormalizeCodegenPlugin } from './codegen/plugin.js';
export { generateNormalizedMetadataForDocs, generateNormalizedMetadata } from './codegen/generateNormalizedMetadata.js';
export { generateNormalizedOperations, generateNormalizedOperation } from './codegen/generateNormalizedOperation.js';

export { NormalizeCodegenPlugin as plugin };

export default {
  plugin: NormalizeCodegenPlugin,
};
