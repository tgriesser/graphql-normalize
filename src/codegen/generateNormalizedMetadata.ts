import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer'
import {
  DocumentNode,
  GraphQLSchema,
  Kind,
  TypeInfo,
  isListType,
  isWrappingType,
  visit,
  print,
  visitWithTypeInfo,
  ArgumentNode,
  FieldNode,
  OperationTypeNode,
  isObjectType,
  isScalarType,
  InlineFragmentNode,
  getNamedType,
  assertObjectType,
  isEnumType,
} from 'graphql'
import type {
  FieldDef,
  FieldMeta,
  NormalizeMetaShape,
  NormalizedDocShape,
  UnionMeta,
  VariableMeta,
} from '../metadataShapes'
import { TypePolicies, getCacheKey } from './getCacheKey'
import { stringifyVariables } from '../stringifyVariables'

export function generateNormalizedMetadata(
  schema: GraphQLSchema,
  document: DocumentNode,
  typePolicies: TypePolicies = {}
) {
  return generateNormalizedMetadataForDocs(schema, [document], typePolicies)[0].meta
}

/**
 * Given an operation and a schema, generates the metadata necessary to
 * write the operation into the cache
 */
export function generateNormalizedMetadataForDocs(
  schema: GraphQLSchema,
  documents: DocumentNode[],
  typePolicies: TypePolicies = {}
): NormalizedDocShape[] {
  let finalDocuments = documents
  if (documents.some((o) => o.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION))) {
    finalDocuments = optimizeDocuments(schema, documents, {
      noLocation: true,
      includeFragments: false,
    })
  }

  const normalizedDocs: NormalizedDocShape[] = []

  const typeInfo = new TypeInfo(schema)
  let normalizedDoc: NormalizeMetaShape = {
    operation: OperationTypeNode.QUERY,
    variables: [],
    fields: [],
  }

  let parentStack: Array<FieldMeta | UnionMeta | NormalizeMetaShape> = []
  let parentFieldDef: FieldMeta | NormalizeMetaShape | UnionMeta = normalizedDoc

  function popStack() {
    const parent = parentStack.pop()
    if (parent) {
      parentFieldDef = parent
    }
  }

  function pushField(field: string | FieldDef) {
    parentFieldDef.fields ??= []
    if (typeof field === 'string' && parentFieldDef.fields.includes(field)) {
      return
    }
    parentFieldDef.fields.push(field)
    if (typeof field !== 'string') {
      parentStack.push(parentFieldDef)
      parentFieldDef = field
    }
  }

  function normalizeArgs(nodes: readonly ArgumentNode[]) {
    let hasVar = false
    const argsDef: Record<string, any> = {}
    for (const arg of nodes) {
      if (arg.value.kind === Kind.VARIABLE) {
        argsDef[`$${arg.value.name.value}`] = arg.name.value
        hasVar = true
      } else {
        argsDef[arg.name.value] = print(arg.value)
      }
    }
    return hasVar ? argsDef : stringifyVariables(argsDef)
  }

  const visitor = visitWithTypeInfo(typeInfo, {
    OperationDefinition: {
      enter(op) {
        normalizedDoc.operation = op.operation
      },
      leave(op) {
        normalizedDocs.push({
          name: op.name?.value ?? 'unnamed',
          meta: normalizedDoc,
        })
        normalizedDoc = {
          operation: OperationTypeNode.QUERY,
          variables: [],
          fields: [],
        }
        parentFieldDef = normalizedDoc
        parentStack = []
      },
    },
    VariableDefinition(node) {
      const obj: VariableMeta = {
        name: node.variable.name.value,
      }
      if (node.defaultValue) {
        obj.defaultValue = print(node.defaultValue)
      }
      normalizedDoc.variables.push(obj)
    },
    Field: {
      enter(node) {
        const { gqlType, listDepth } = unpackType(typeInfo)

        // If this is a boring ol scalar / enum, we push it as a string
        if (isSimpleField(node, typeInfo)) {
          pushField(node.name.value)
          return
        }

        const fieldMeta: FieldMeta = { name: node.name.value }
        pushField(fieldMeta)

        if (listDepth) fieldMeta.list = listDepth
        if (node.alias?.value) fieldMeta.alias = node.alias.value

        if (isObjectType(gqlType)) {
          const cacheKey = getCacheKey(typePolicies, gqlType)
          if (cacheKey) {
            fieldMeta.cacheKey = cacheKey
          }
          if (gqlType.name === schema.getQueryType()?.name) {
            fieldMeta.cacheKey = '$root'
          }
        }
        if (node.arguments?.length) {
          fieldMeta.args = normalizeArgs(node.arguments)
        }

        const skipDirective = node.directives?.find((d) => d.name.value === 'skip')
        const includeDirective = node.directives?.find((d) => d.name.value === 'include')
        if (skipDirective) {
          fieldMeta.skip = normalizeArgs(skipDirective.arguments ?? [])
        }
        if (includeDirective) {
          fieldMeta.include = normalizeArgs(includeDirective.arguments ?? [])
        }
      },
      leave(node) {
        // Pop things back into place
        if (!isSimpleField(node, typeInfo)) {
          popStack()
        }
      },
    },
    InlineFragment: {
      enter(node) {
        if (
          node.typeCondition?.name.value &&
          isConcreteAbstract(node, typeInfo, schema) &&
          parentIsFieldMeta(parentFieldDef)
        ) {
          const unionType: UnionMeta = {
            fields: [],
          }
          const outputType = getNamedType(typeInfo.getType())
          const cacheKey = getCacheKey(typePolicies, assertObjectType(outputType))
          if (cacheKey) {
            unionType.cacheKey = cacheKey
          }

          parentFieldDef.possible ??= {}
          parentFieldDef.possible[node.typeCondition?.name.value] = unionType
          parentStack.push(parentFieldDef)
          parentFieldDef = unionType
        }
      },
      leave(node) {
        if (node.typeCondition?.name.value && isConcreteAbstract(node, typeInfo, schema)) {
          popStack()
        }
      },
    },
    FragmentDefinition() {
      throw noFragments()
    },
  })

  for (const doc of finalDocuments) {
    visit(doc, visitor)
  }

  return normalizedDocs
}

function unpackType(typeInfo: TypeInfo) {
  let listDepth = 0
  // We want to unwrap any list types, and make those the containing
  let gqlType = typeInfo.getType()
  while (isWrappingType(gqlType)) {
    if (isListType(gqlType)) {
      ++listDepth
    }
    gqlType = gqlType.ofType
  }
  return { gqlType, listDepth }
}

function isSimpleField(node: FieldNode, typeInfo: TypeInfo) {
  const { gqlType, listDepth } = unpackType(typeInfo)
  return (
    (isScalarType(gqlType) || isEnumType(gqlType)) &&
    !listDepth &&
    !node.alias &&
    !node.arguments?.length &&
    !node.directives?.some((d) => d.name.value === 'skip' || d.name.value === 'includes')
  )
}

function noFragments() {
  return new Error('Fragment definitions have been stripped')
}

function isConcreteAbstract(node: InlineFragmentNode, typeInfo: TypeInfo, schema: GraphQLSchema) {
  if (node.typeCondition?.name) {
    const parentTypeName = typeInfo.getParentType()?.name
    if (node.typeCondition.name.value !== parentTypeName) {
      return isObjectType(schema.getType(node.typeCondition?.name.value))
    }
  }
  return false
}

function parentIsFieldMeta(parent: FieldMeta | NormalizeMetaShape | UnionMeta): parent is FieldMeta {
  return 'name' in parent
}
