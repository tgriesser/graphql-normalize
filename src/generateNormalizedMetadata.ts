import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import {
  DocumentNode,
  GraphQLSchema,
  Kind,
  TypeInfo,
  isAbstractType,
  isListType,
  isWrappingType,
  visit,
  print,
  visitWithTypeInfo,
  ArgumentNode,
  FieldNode,
  SelectionSetNode,
  OperationTypeNode,
  isObjectType,
} from 'graphql';
import stableStringify from 'fast-json-stable-stringify';
import type { FieldDef, FieldMeta, NormalizeMeta, NormalizedDoc, VariableMeta } from './extensionShape';
import { TypePolicies, getCacheKey } from './typePolicies';

/**
 * Given an operation and a schema, generates the metadata necessary to
 * write the operation into the cache
 */
export function generateNormalizedMetadata(
  schema: GraphQLSchema,
  operation: DocumentNode,
  typePolicies: TypePolicies = {}
): NormalizedDoc {
  let finalOperation = operation;
  if (operation.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION)) {
    finalOperation = optimizeDocuments(schema, [operation], {
      noLocation: true,
    })[0];
  }

  const typeInfo = new TypeInfo(schema);
  const normalizedDoc: NormalizedDoc = {
    operation: OperationTypeNode.QUERY,
    variables: [],
    selectionSet: {},
  };

  const fieldPath: any[] = [];

  function setPath<K extends keyof NormalizeMeta>(key: K, value: NormalizeMeta[K]) {
    normalizedDoc.selectionSet[fieldPath.join('.')] ??= {};
    normalizedDoc.selectionSet[fieldPath.join('.')][key] = value;
  }
  function normalizeArgs(nodes: readonly ArgumentNode[]) {
    let hasVar = false;
    const argsDef: Record<string, any> = {};
    for (const arg of nodes) {
      if (arg.value.kind === Kind.VARIABLE) {
        argsDef[`$${arg.name.value}`] = arg.name.value;
        hasVar = true;
      } else {
        argsDef[arg.name.value] = print(arg.value);
      }
    }
    return hasVar ? argsDef : stableStringify(argsDef);
  }
  function normalizeField(node: FieldNode) {
    if (node.alias || node.arguments?.length) {
      const field: FieldMeta = {
        name: node.name.value,
      };
      if (node.alias) field.alias = node.alias.value;
      if (node.arguments?.length) field.args = normalizeArgs(node.arguments);
      return field;
    }
    return node.name.value;
  }
  function addSelectionFields(node: SelectionSetNode) {
    const fields: FieldDef[] = [];
    for (const sel of node.selections) {
      if (sel.kind === Kind.FRAGMENT_SPREAD) {
        throw noFragments();
      }
      if (sel.kind === Kind.FIELD) {
        fields.push(normalizeField(sel));
      }
      // If we have an inline fragment
      if (sel.kind === Kind.INLINE_FRAGMENT) {
        //
      }
    }
    setPath('fields', fields);
  }

  const visitor = visitWithTypeInfo(typeInfo, {
    OperationDefinition(op) {
      normalizedDoc.operation = op.operation;
    },
    VariableDefinition(node) {
      const obj: VariableMeta = {
        name: node.variable.name.value,
      };
      if (node.defaultValue) {
        obj.defaultValue = print(node.defaultValue);
      }
      normalizedDoc.variables.push(obj);
    },
    Field: {
      enter(node) {
        if (node.alias?.value) {
          fieldPath.push(node.alias.value);
          setPath('aliasField', node.name.value);
        } else {
          fieldPath.push(node.name.value);
        }
        let gqlType = typeInfo.getType();
        while (isWrappingType(gqlType)) {
          if (isListType(gqlType)) {
            setPath('list', true);
            fieldPath.push('$idx');
          }
          gqlType = gqlType.ofType;
        }
        if (isAbstractType(gqlType)) {
          setPath(
            'possible',
            schema.getPossibleTypes(gqlType).map((t) => t.name)
          );
        }
        if (isObjectType(gqlType)) {
          const cacheKey = getCacheKey(typePolicies, gqlType);
          if (cacheKey) {
            setPath('cacheKey', cacheKey);
          }
        }
        if (node.arguments?.length) {
          setPath('args', normalizeArgs(node.arguments));
        }
        if (node.selectionSet) {
          addSelectionFields(node.selectionSet);
        }
      },
      leave(node) {
        fieldPath.pop();
        let gqlType = typeInfo.getType();
        while (isWrappingType(gqlType)) {
          if (isListType(gqlType)) {
            fieldPath.pop();
          }
          gqlType = gqlType.ofType;
        }
      },
    },
    InlineFragment: {
      enter(node) {
        if (node.typeCondition?.name) {
          fieldPath.push(`$type_${node.typeCondition?.name.value}`);
          addSelectionFields(node.selectionSet);
        }
      },
      leave(node) {
        if (node.typeCondition?.name.value) {
          fieldPath.pop();
        }
      },
    },
    FragmentDefinition() {
      throw noFragments();
    },
  });

  visit(finalOperation, visitor);

  return normalizedDoc;
}

function noFragments() {
  return new Error('Fragment definitions have been stripped');
}
