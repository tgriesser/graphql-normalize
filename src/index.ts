import { graphqlNormalize } from './graphqlNormalize.js';
export { stringifyVariables } from './stringifyVariables.js';
export { MissingFieldError } from './errors.js';
export type { graphqlNormalizeOptions, NormalizeMetaShape } from './graphqlNormalize.js';
export { graphqlNormalize };
export default graphqlNormalize;
