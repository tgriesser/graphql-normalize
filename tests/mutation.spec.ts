import { beforeEach, describe, expect, it } from 'vitest'
import { generateNormalizedMetadata } from '../src/codegen/generateNormalizedMetadata'
import { schema } from './fixtures/schema'
import { mutation1Doc, operation3Doc } from './fixtures/ops'
import { NormalizeMetaShape, graphqlNormalize } from '../src'
import { ExecutionResult, execute } from 'graphql'
import { generateNormalizedOperation } from '../src/codegen'
import _ from 'lodash'

describe('mutation', () => {
  let meta: NormalizeMetaShape
  let variableValues = {}
  let result: ExecutionResult
  let cache: Record<string, any>

  beforeEach(async () => {
    meta = generateNormalizedMetadata(schema, operation3Doc)
    variableValues = {}
    result = await execute({
      schema,
      variableValues,
      document: generateNormalizedOperation(schema, operation3Doc),
    })
    cache = {} as const
  })

  it('should generate metadata for a mutation', () => {
    expect(generateNormalizedMetadata(schema, mutation1Doc)).toMatchSnapshot()
  })

  it('syncs the mutation result with the cache', async () => {
    const obj = graphqlNormalize({
      action: 'write',
      meta,
      variableValues,
      operationResult: result,
      cache,
      isEqual: _.isEqual,
      //
    })

    expect({
      added: obj.added,
      modified: obj.modified,
    }).toEqual({ added: 44, modified: 0 })

    expect(cache).toMatchSnapshot()

    const {
      added,
      modified,
      result: result2,
      cache: updatedCache,
    } = graphqlNormalize({
      action: 'write',
      meta: generateNormalizedMetadata(schema, mutation1Doc),
      variableValues,
      operationResult: await execute({
        schema,
        variableValues,
        document: generateNormalizedOperation(schema, mutation1Doc),
      }),
      cache,
      isEqual: _.isEqual,
    })

    expect({ added, modified }).toEqual({ added: 21, modified: 10 })

    expect(updatedCache).toMatchSnapshot()

    expect(result2).toMatchSnapshot()
  })
})
