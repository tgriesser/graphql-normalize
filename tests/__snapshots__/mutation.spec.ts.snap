// Vitest Snapshot v1

exports[`mutation > should generate metadata for a mutation 1`] = `
{
  "fields": [
    {
      "fields": [
        {
          "cacheKey": "Post:id",
          "fields": [
            "id",
            "title",
          ],
          "name": "post",
        },
        {
          "cacheKey": "$root",
          "fields": [
            {
              "alias": "first10Posts",
              "args": "{\\"first\\":\\"10\\"}",
              "fields": [
                {
                  "cacheKey": "Post:id",
                  "fields": [
                    "id",
                    "title",
                  ],
                  "list": 1,
                  "name": "nodes",
                },
              ],
              "name": "posts",
            },
          ],
          "name": "query",
        },
      ],
      "name": "addPost",
    },
  ],
  "operation": "mutation",
  "variables": [],
}
`;

exports[`mutation > syncs the mutation result with the cache 1`] = `
{
  "Post:UG9zdDo0": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDo0",
    },
    "title": {
      "$": "Some Blog post 4",
    },
  },
  "Post:UG9zdDo1": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDo1",
    },
    "title": {
      "$": "Some Blog post 5",
    },
  },
  "Post:UG9zdDo2": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDo2",
    },
    "title": {
      "$": "Some Blog post 6",
    },
  },
  "Post:UG9zdDo3": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDo3",
    },
    "title": {
      "$": "Some Blog post 7",
    },
  },
  "Post:UG9zdDo4": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDo4",
    },
    "title": {
      "$": "Some Blog post 8",
    },
  },
  "Post:UG9zdDo5": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDo5",
    },
    "title": {
      "$": "Some Blog post 9",
    },
  },
  "Post:UG9zdDox": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDox",
    },
    "title": {
      "$": "Some Blog post 1",
    },
  },
  "Post:UG9zdDoxMA==": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDoxMA==",
    },
    "title": {
      "$": "Some Blog post 10",
    },
  },
  "Post:UG9zdDoy": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDoy",
    },
    "title": {
      "$": "Some Blog post 2",
    },
  },
  "Post:UG9zdDoz": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDoz",
    },
    "title": {
      "$": "Some Blog post 3",
    },
  },
  "posts": {
    "{\\"first\\":\\"10\\"}": {
      "nodes": {
        "$": [
          {
            "$ref": "Post:UG9zdDox",
          },
          {
            "$ref": "Post:UG9zdDoy",
          },
          {
            "$ref": "Post:UG9zdDoz",
          },
          {
            "$ref": "Post:UG9zdDo0",
          },
          {
            "$ref": "Post:UG9zdDo1",
          },
          {
            "$ref": "Post:UG9zdDo2",
          },
          {
            "$ref": "Post:UG9zdDo3",
          },
          {
            "$ref": "Post:UG9zdDo4",
          },
          {
            "$ref": "Post:UG9zdDo5",
          },
          {
            "$ref": "Post:UG9zdDoxMA==",
          },
        ],
      },
    },
  },
}
`;

exports[`mutation > syncs the mutation result with the cache 2`] = `
{
  "Post:UG9zdDo0": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDo0",
    },
    "title": {
      "$": "Some Blog post 4",
    },
  },
  "Post:UG9zdDo1": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDo1",
    },
    "title": {
      "$": "Some Blog post 5",
    },
  },
  "Post:UG9zdDo2": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDo2",
    },
    "title": {
      "$": "Some Blog post 6",
    },
  },
  "Post:UG9zdDo3": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDo3",
    },
    "title": {
      "$": "Some Blog post 7",
    },
  },
  "Post:UG9zdDo4": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDo4",
    },
    "title": {
      "$": "Some Blog post 8",
    },
  },
  "Post:UG9zdDo5": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDo5",
    },
    "title": {
      "$": "Some Blog post 9",
    },
  },
  "Post:UG9zdDox": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDox",
    },
    "title": {
      "$": "Some Blog post 1",
    },
  },
  "Post:UG9zdDoxMA==": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDoxMA==",
    },
    "title": {
      "$": "Some Blog post 10",
    },
  },
  "Post:UG9zdDoxMDI=": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDoxMDI=",
    },
    "title": {
      "$": "Added post",
    },
  },
  "Post:UG9zdDoy": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDoy",
    },
    "title": {
      "$": "Some Blog post 2",
    },
  },
  "Post:UG9zdDoz": {
    "__typename": "Post",
    "id": {
      "$": "UG9zdDoz",
    },
    "title": {
      "$": "Some Blog post 3",
    },
  },
  "posts": {
    "{\\"first\\":\\"10\\"}": {
      "nodes": {
        "$": [
          {
            "$ref": "Post:UG9zdDoxMDI=",
          },
          {
            "$ref": "Post:UG9zdDox",
          },
          {
            "$ref": "Post:UG9zdDoy",
          },
          {
            "$ref": "Post:UG9zdDoz",
          },
          {
            "$ref": "Post:UG9zdDo0",
          },
          {
            "$ref": "Post:UG9zdDo1",
          },
          {
            "$ref": "Post:UG9zdDo2",
          },
          {
            "$ref": "Post:UG9zdDo3",
          },
          {
            "$ref": "Post:UG9zdDo4",
          },
          {
            "$ref": "Post:UG9zdDo5",
          },
        ],
      },
    },
  },
}
`;

exports[`mutation > syncs the mutation result with the cache 3`] = `
{
  "addPost": {
    "post": {
      "id": "UG9zdDoxMDI=",
      "title": "Added post",
    },
    "query": {
      "first10Posts": {
        "nodes": [
          {
            "id": "UG9zdDoxMDI=",
            "title": "Added post",
          },
          {
            "id": "UG9zdDox",
            "title": "Some Blog post 1",
          },
          {
            "id": "UG9zdDoy",
            "title": "Some Blog post 2",
          },
          {
            "id": "UG9zdDoz",
            "title": "Some Blog post 3",
          },
          {
            "id": "UG9zdDo0",
            "title": "Some Blog post 4",
          },
          {
            "id": "UG9zdDo1",
            "title": "Some Blog post 5",
          },
          {
            "id": "UG9zdDo2",
            "title": "Some Blog post 6",
          },
          {
            "id": "UG9zdDo3",
            "title": "Some Blog post 7",
          },
          {
            "id": "UG9zdDo4",
            "title": "Some Blog post 8",
          },
          {
            "id": "UG9zdDo5",
            "title": "Some Blog post 9",
          },
        ],
      },
    },
  },
}
`;
